evm_loader/program/src/entrypoint.rs

```diff
@@ -731,51 +653,32 @@ fn invoke_on_return<'a>(
     Ok(())
 }
 
-fn check_from_or_signer<'a>(
-    program_id: &Pubkey,
-    caller_opt: Option<&SolidityAccount<'a>>,
-    caller_info: &'a AccountInfo<'a>,
-    signer_info: &'a AccountInfo<'a>,
-    from_info: Option<(H160, u64)>,
-) ->  ProgramResult
+fn check_ethereum_authority<'a>(
+   sender: &SolidityAccount<'a>,
+   recovered_address: &H160,
+   trx_nonce: u64,
+   chain_id: &U256,
+) -> ProgramResult
 {
-    if caller_info.owner == program_id {
-        if caller_opt.is_some() {
-            let caller = caller_opt.unwrap();
-
-            let caller_signer = caller.get_signer();
-            let caller_ether = caller.get_ether();
-            let caller_nonce = caller.get_nonce();
-
-            if from_info.is_none() {
-                if caller_signer != *signer_info.key || !signer_info.is_signer {
-                    debug_print!("Add valid account signer");
-                    debug_print!("   caller signer: {}", &caller_signer.to_string());
-                    debug_print!("   signer pubkey: {}", &signer_info.key.to_string());
-                    debug_print!("is signer signer: {}", &signer_info.is_signer.to_string());
-
-                    return Err(ProgramError::InvalidArgument);
-                }
-            } else {
-                let (from, nonce) = from_info.unwrap();
-                if caller_ether != from {
-                    debug_print!("Invalid caller account");
-                    debug_print!("   caller addres: {}", &caller_ether.to_string());
-                    debug_print!("     from addres: {}", &from.to_string());
+    if sender.get_ether() != *recovered_address {
+        debug_print!("Invalid sender: actual {}, recovered {}",
+                &sender.get_ether().to_string(),
+                &recovered_address.to_string());
+        return Err(ProgramError::InvalidArgument);
+    }
 
-                    return Err(ProgramError::InvalidArgument);
-                }
-                if caller_nonce != nonce {
-                    debug_print!("Invalid Ethereum transaction nonce");
-                    debug_print!("     tx nonce: {}", &nonce.to_string());
-                    debug_print!("    acc nonce: {}", &caller_nonce.to_string());
+    if sender.get_nonce() != trx_nonce {
+        debug_print!("Invalid Ethereum transaction nonce: acc {}, trx {}",
+                &sender.get_nonce().to_string(),
+                &trx_nonce.to_string());
+        return Err(ProgramError::InvalidArgument);
+    }
 
-                    return Err(ProgramError::InvalidArgument);
-                }
-            }
-        } else {
-            return Err(ProgramError::InvalidArgument);
-        }
+    if SolanaBackend::<ProgramAccountStorage>::chain_id() != *chain_id {
+        debug_print!("Invalid chain_id: actual {}, expected {}",
+                &chain_id.to_string(),
+                &SolanaBackend::<ProgramAccountStorage>::chain_id().to_string());
```

Comment:
`&chain_id.to_string()` can be simplified to `chain_id`.
Also check other debug_print macro

Reply:
Fixed